### PR TITLE
[JENKINS-26574] Allow distinct template for matrix head build

### PR DESCRIPTION
**Pull request by lavaliere**
[JENKINS-26574](https://issues.jenkins-ci.org/browse/JENKINS-26574) - Added support for Matrix projects by adding an extra field and hiding this field if the project type is not Matrix. There are 3 cases I can think of when it comes to these fields:

* template set, matrix empty (presumably != matrix build, ignore matrix field values)
* matrix set, template empty (presumably matrix build with child builds taking the same name pattern as the parent build)
* template and matrix set - use case that a CloudBees customer was asking for. Allows Matrix build name to be set explicitly and separately from spawned builds in the case where spawned builds take their name from a build-generated file.

I also added a test cases to each method to check that both case 2 and case 3 above are being set correctly.

### CODE REVIEW

**Comment by lavaliere** [comment]
@reviewbybees

**Comment by daniel-beck** [comment]
Commits should probably be squashed.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 1]
```diff
@@ -26,48 +28,106 @@
...
+					e.printStackTrace();
+				}
+                return false;
             }
 
             @Override
             public boolean endBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.endBuild();
+            	try{
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Inconsistent coding style (see line 108).

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 2]
```diff
@@ -26,48 +28,106 @@
...
     public MatrixAggregator createAggregator(MatrixBuild build, Launcher launcher, BuildListener listener) {
         return new MatrixAggregator(build,launcher,listener) {
             @Override
             public boolean startBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.startBuild();
+                try {
+					setDisplayName(build,listener);
+					started = true;
+	                return super.startBuild();
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Indentation is a mess here.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 3]
```diff
@@ -5,11 +5,13 @@
 import hudson.matrix.MatrixAggregatable;
 import hudson.matrix.MatrixAggregator;
 import hudson.matrix.MatrixBuild;
+import hudson.matrix.MatrixProject;
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: While the plugin is currently on 1.480.x, what happens to this dependency when that baseline gets raised? Since Matrix Plugin has been detached from core a while back, is there a way for this PR to be implemented that allows the Matrix Plugin dependency to be made optional?

File: src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java [hunk 4]
```diff
@@ -1,72 +1,156 @@
-package org.jenkinsci.plugins.buildnamesetter;
```
Comment (daniel-beck) [inline src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java]: I don't know what you did here (line breaks change from Windows/Unix to Unix/Windows), but the diff shouldn't look like this.

File: src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/help-template.jelly [hunk 5]
```diff
@@ -3,4 +3,13 @@
     setting what name new build gets. This field can contain the following macros:
 
     <help xmlns="/lib/token-macro" />
+
+    For multi-configuration projects: if the Matrix Build Name is set and the Build 
+    Name field is not set, then the head build and child builds will all use the Matrix Build 
+    Name's pattern.
+
```
Comment (daniel-beck) [inline src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/help-template.jelly]: Not actually a paragraph break in the rendered HTML output.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 6]
```diff
@@ -26,48 +28,106 @@
...
+    private boolean started = false;
+    
     @DataBoundConstructor
+    public BuildNameSetter(String template, String matrixTemplate) {
+        this.template = template;
+        this.matrixTemplate = matrixTemplate;
+        matches = template.equals(matrixTemplate);
+        
+        if(matches || matrixTemplate.length()>0 && template.length()==0){
+        	scenario = 0;
```
Comment (recampbell) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Please use 4 spaces instead of tabs when indenting.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 7]
```diff
@@ -26,48 +28,106 @@
...
+            		setDisplayName(build,listener);
+            		return super.endBuild();
+            	}catch(Exception e){
+            		e.printStackTrace();
+            	}
+            	return false;
             }
         };
     }
+
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Try to avoid unnecessary whitespace changes in existing code.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java (hunk 2 above)
Comment (recampbell) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: This is because tabs are used instead of spaces.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 8]
```diff
@@ -26,48 +28,106 @@
...
+    public BuildNameSetter(String template, String matrixTemplate) {
+        this.template = template;
+        this.matrixTemplate = matrixTemplate;
+        matches = template.equals(matrixTemplate);
+        
+        if(matches || matrixTemplate.length()>0 && template.length()==0){
+        	scenario = 0;
+        }else if (matrixTemplate.length()>0 && template.length()>0){
+        	scenario = 1;
+        }else if (matrixTemplate.length()==0 && template.length()>0){
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: The plugin's current coding style is to add spaces before open curly braces and after closing curly braces, so this should probably be made consistent.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java (hunk 3 above)
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Correction: This change will probably break the plugin for anyone having disabled Matrix Project.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 9]
```diff
@@ -26,48 +28,106 @@
...
     @Override
     public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
-        setDisplayName(build, listener);
-
-        return new Environment() {
+        try{
+        	//not checking before setting name because this check is done via Jelly call to isMatrix()
+        	setDisplayName(build, listener);
+        }catch(Exception e){
+        	e.printStackTrace();
```
Comment (christ66) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Should be `e.printStackTrace(listener.getLogger())` otherwise the stack trace and error message are lost to standard out/err. Also the catch block should be changed to only catch `IllegalArgumentException`.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 10]
```diff
@@ -26,48 +28,106 @@
...
+	    				build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+	    			}else{
+	    				build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+	    			}
+    				return;
+    		case 2:
+	    			build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+    				return;
+    		case -1:
+    				throw new Exception("No valid values set for build names!");
```
Comment (christ66) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Should probably not be throwing a new `Exception`, but instead `java.lang.IllegalArgumentException`. Then the `private void setDisplayName(AbstractBuild build, BuildListener listener) throws Exception` should be changed to throw `IllegalArgumentException`.

**Comment by stephenc** [comment]
@daniel-beck I disagree, as long as the merge is not fast-forwarded. If you are going to fast-forward the merge then yes the commits *must* be squashed, but otherwise it can be good to see the thought processes of the developer

**Comment by jglick** [comment]
> If you are going to fast-forward the merge

Note that the _Merge pull request_ button passes the equivalent of `--no-ff` so this is usually not a concern. I would also discourage use of history rewriting (except in special cases such as large binaries being inadvertently committed).

**Comment by lavaliere** [comment]
@reviewbybees - attempted to make all of the selected changes re: styling (switched from Eclipse to NetBeans), exceptions, and making the matrix-project optional

File: pom.xml [hunk 11]
```diff
@@ -3,7 +3,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.480.3</version>
+    <version>1.599</version>
```
Comment (stephenc) [inline pom.xml]: Really? can you not live with even the most recent LTS?

I would think you could even go as far back as 1.565.1

**Comment by lavaliere** [comment]
@stephenc Point taken. 

I changed the required version to 1.561, which is the minimum required for the version of the Matrix plugin whose methods we call in this code. I had changed it to 1.599 on the suggestion of Jesse's NetBeans Jenkins plugin. 

Perhaps leaving 1.480 as the required version would be preferred though, since the Matrix builds calls are conditional on the plugin being installed at all...?

**Comment by alecharp** [comment]
@lavaliere - I think 1.561 is good because of the Matrix plugin version used (1.4). Letting the 1.480.x version could lead to other issues which would be complicated to fix on 1.480.x base.

**Comment by daniel-beck** [comment]
It appears the current plugin already has a dependency on matrix projects (as it currently just sets the parent build's description to the same string), I just misread the earlier change. Sorry!

Making the dependency optional is probably misleading; what happens when you run this with Matrix Project Plugin disabled? I'd expect it to fail to load the publisher as it is (and was) `MatrixAggregatable`.

(This is what happens when dependencies to detached plugins are implied, so raising the base version to 1.561-1.565.x would make the dependency explicit at least, without excluding a majority of users.)

Using LTS versions as plugin base versions is preferred, see https://wiki.jenkins-ci.org/display/JENKINS/Plugin+Report+Card.

**Comment by ndeloof** [comment]
see https://github.com/jenkinsci/artifactdeployer-plugin/pull/8 as an attempt to make matrix-plugin an optional dependency but address the `MatrixAggregatable` issue

**Comment by jglick** [comment]
> I had changed it to 1.599 on the suggestion of Jesse's NetBeans Jenkins plugin.

For the record: my plugin does not suggest changes to the baseline version. The generic Maven support does suggest using the latest release version of a parent POM if a newer one is known from the Maven index. In this case the suggestion should generally be ignored.

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java [hunk 12]
```diff
@@ -26,50 +28,106 @@
...
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
         };
     }
+   
 
-    @Extension
+    @Extension(optional = true)
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: As noted elsewhere, this is just wrong; you cannot make the dependency on `matrix-project` optional this way. Currently I think there is _no_ reasonable way to make it optional.

**Comment by stephenc** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/jenkinsci/build-name-setter-plugin/4)

Reviewed files:
- pom.xml @ r3
- src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java @ r2
- src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/config.jelly @ r1
- src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/help-template.jelly @ r1
- src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java @ r1

---

Lots of excessive whitespace changes

---

<sup>**[src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java, line 37 \[r3\]](https://reviewable.io:443/reviews/jenkinsci/build-name-setter-plugin/4#-JlMu_8O6voigHK1JsvJ)** ([raw file](https://github.com/jenkinsci/build-name-setter-plugin/blob/fe06a7abae8571f8426c645fb8e091c9289ee35b/src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java#L37)):</sup>
gratuitous whitespace changes

---

File: src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java (hunk 12 above)
Comment (lavaliere) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Okay. I got that tag from the OSS wiki [on plugin dependencies](https://wiki.jenkins-ci.org/display/JENKINS/Dependencies+among+plugins): 

> You can mark your extension as an optional extension by changing the @Extension annotation to @Extension(optional = true) and Jenkins don't log any class loading errors when reading it.

I'll look through the options [on this page](https://wiki.jenkins-ci.org/display/JENKINS/Tips+for+optional+dependencies) instead.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: Yes `Extension.optional` exists and yes it sometimes works, but you have to know exactly what you are doing. Basically all it does is tell Jenkins that if _when loading the extension_ a `LinkageError` (typically `NoClassDefFoundError`) is thrown, to quietly skip the extension rather than print a big stack trace. For some simple cases, like extensions that just implement an SPI in some other plugin, that is enough. But making the dependency optional here is much more complicated than that.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java]: As noted above, I do not think it is feasible for the dependency to be in any way optional currently. Would really require a new API in `matrix-project`. Anyway optional dependencies are always a headache; prefer to just keep it as a hard dep.

**Comment by stephenc** [comment]
:bug: as there are a lot of outstanding comments remaining here.

**Comment by ebrahim-moshaya** [comment]
This update would be really useful!! Any chance for a merge?

**Comment by lavaliere** [comment]
Sorry, it's been so long since I wrote this that I had assumed it had been merged. I'll review the comments and update my PR this weekend.

**Comment by ebrahim-moshaya** [comment]
@lavaliere thanks for the update, look forward to this getting merged

**Comment by apemberton** [comment]
Has this been merged?

**Comment by oleg-nenashev** [comment]
@apemberton 
The PR is not ready for merge, CC @lavaliere

**Comment by Le0Michine** [comment]
Is the PR alive?

**Comment by daniel-beck** [comment]
@lavaliere Ping!

**Comment by Le0Michine** [comment]
@lavaliere please respond, I'm going to close pull request if it is not needed more and doesn't ready to merge

**Comment by oleg-nenashev** [comment]
I've pinged @lavaliere directly. If there is no response within 24 hours, feel free to close it (you are eligible to do it as a maintainer). BTW I recommend to just add the "needs-fix" label and forget about it. If somebody wants to finish it, he will be able to use the current stuff as a source
